### ISSUE TITLE
Unexpected behavior: interfaces/classes with the property "name" are assignable to function type

### ISSUE THREAD

**Issue by Tiagoperes** [microsoft/TypeScript#48375]
# Bug Report

### 🔎 Search Terms
interface class property name assigned function

### 🕗 Version & Regression Information

I'm not sure if this is a bug and it's the first time I observed the issue. Through the playground I confirmed this happens in every version of TS.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.6.2#code/JYOwLgpgTgZghgYwgAgMIHsC2AHdILjIDeAUMssACYD8AXMgM5hSgDmANGctlOttGGAQGdZACUICdFEoAeJixAdkcEAE8AfJ3Ig4mCPQVttyXfobZEEUUaUmEAC2AAbSlAKiMOPATABtAF1OAF8SEgB6cOQASWRKKlN0MGQIAA9+BGSwJwZkMHRkACMURCRsSEoSKRAmZClvfHB6L1xG5IBeZAAKAEpkdo1kAFYSIA)

### 💻 Code

The problem is: I use classes and Interfaces to dictate the structure of Objects. I'd never expect to receive a function. If I intend to receive a function, I write something like `(argument: ArgumentType) => ResponseType`.

The code below doesn't raise any errors, because a function does have the property "name". But I really didn't expect this. Is this the correct behavior? Should I add a new Attribute to my interface/class just so a function can't be assigned to it?

```typescript
interface Component {
  id?: string,
  properties?: Record<string, any>,
  name: string,
  namespace?: string,
  children?: Component[],
}

// I did not expect this to be accepted
const component: Component = () => 5
```

### 🙁 Actual behavior

It accepted my code

### 🙂 Expected behavior

I expected it to fail due to type constraints

**Comment by kaya3** [comment]
This looks like a pitfall but I think it is the intended behaviour, since in general interfaces can be used for functions. A workaround is to add an optional property like `call?: never` so that functions (which do have a `call` property) aren't assignable.

[Playground Link](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgMIHsC2AHdILjIDeAUMssACYD8AXMgM5hSgDmANGctlOttGGAQGdZACUICdFEoAeJixAdkcEAE8AfJ3Ig4mCPQVttyXfobZEEUUaUmEAC2AAbSlAKiMOPATABtAF1OLgB6ACowrmQw5AAVJwZuXn4oMDUKRMpJZzh3SkZ0ZDAHODBkGABXEARBPAZ2ZAB3J0dkSnQomJKANxQ4ZAADBDhnZwGkvgE1BtyUEHQyuAYGYFZdACNnFDBCr1x8cAA6TpCuYdHRfF6oTgBfEhIpECZkKW8DsHo9n0IAXmQABQASmQvw0yAArCQgA)

**Comment by RyanCavanaugh** [comment]
> The code below doesn't raise any errors, because a function does have the property "name". But I really didn't expect this

There's no ad-hoc rule that says functions can't be assignment sources, and indeed doing so would break other scenarios (since class constructors are also functions):
```ts
class Something {
  static foo: string
}
// No problems here
let s: { foo: string } = Something;
```

**Comment by Tiagoperes** [comment]
@kaya3 yeah, I thought this could be the case. It's a very unintuitive aspect of Typescript though. Pretty weird to create a class, expect an instance of it and end up with a function.

@RyanCavanaugh 
Great example. You're right, Javascript can be very weird sometimes, the following is totally correct:
```typescript
function something() {
  // ...
}

something.foo = 'bar'
console.log(something.foo)
```
`something` is a function, but it can also behave like an object and there's nothing in the JS spec prohibiting me from doing this. Since `typeof class Foo {}` is also "function", we can't do much here.

I bet most people would never expect this behavior, but I don't think TS can do anything to change this.

Thank you both very much for your perspectives on the subject. I didn't want too, but I'll add `call: never` to my interface (it feels like a hack =p).
